### PR TITLE
Moved the HphaSchema implementations to a base template classoved the HphaSchema implementations to a base template class

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
This allows removal of the includes of the HphaSchema.cpp files from within the HphaSchemaErrorDetection.cpp file
For the other cpp include of MallocSchema.cpp, it turns out it was never
used anywhere within the HphaSchemaErrorDetection test, so it has been removed.

This allows removal of  the using namespace calls for ::AZStd and ::AZ from the Internal namespace in the HphaSchemeaSchemaErrorDetection.cpp.

For example the `initializer_list` class is aliased in the AZStd namespace. So when `initializer_list` is specified instead of `AZStd::initializer_list`, it would cause a compilation issue because the `Internal::AZStd::initializer_list` symbol is not available because it wasn't aliased inside of the Internal namespace.

Pretty much the issues that occurs with including a cpp file with a namespace is as follows

1. At global namespace scope a file let's say such as <AzCore/base.h> is included.
   That aliases `std::initializer_list` into ::AZStd namespace
2. Later on in the Internal namespace HphaSchema.cpp was included, which includes <AzCore/std/containers/unordered_map.h> for the first time.
  That unordered_map.h would try to include <AzCore/base.h>, but it can't because it was already included outside of the Internal namespace
3. The unordered_map class implementation occurs in the AZStd namespace(which is now Internal::AZStd, because its header was indirectly included within the Internal namespace) references the symbol the `initializer_list` without AZStd in front of it
4. Because there are symbols with the names of `Internal::AZStd::initializer_list`, `Internal::initializer_list` or `::initializer_list`, the compilation fails.

TLDR: The inclusion of a cpp file with other includes inside of a namespace, was putting the onus on the included files to deal with an issue of fully qualifying names, when the issue was caused by the inclusion of the cpp file.

Signed-off-by: lumberyard-employee-dm <56135373+lumberyard-employee-dm@users.noreply.github.com>

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/Memory/HphaSchema.h [hunk 1]
```diff
@@ -5,91 +5,17 @@
...
+#pragma once
 
 #include <AzCore/Memory/Memory.h>
 #include <AzCore/std/typetraits/aligned_storage.h>
 
+ // The HphaSchemaErrorDetection.cpp file test the HphaSchema
+ // implementation with the DEBUG_ALLOCATOR define set to true
+ // In order a second implementation of this class to exist
+ // without polluting the test with using namespace ::AZStd/::AZ
+ // The interface is placed into an inl file with no includes
```
Comment (lawsonamzn) [inline Code/Framework/AzCore/AzCore/Memory/HphaSchema.h:17]: This could use a proofread.  Is it c opied from somewhere?

File: Code/Framework/AzCore/AzCore/Memory/MallocSchema.cpp [hunk 2]
```diff
@@ -11,147 +11,12 @@
...
-        uint32_t offset;
-        uint32_t size;
-    };
-} // namespace AZ::Internal
-
+ // The HphaSchemaErrorDetection.cpp file test the MallocSchema
+ // implementation with the DEBUG_ALLOCATOR define set to true
+ // In order a second implementation of this class to exist
+ // without polluting the test with using namespace ::AZStd/::AZ
+ // The implementation is placed into an inl file with no includes
```
Comment (lawsonamzn) [inline Code/Framework/AzCore/AzCore/Memory/MallocSchema.cpp:18]: A copy of the previous text with same errors?

File: Code/Framework/AzCore/AzCore/Memory/HphaSchema.h (hunk 1 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Memory/HphaSchema.h:17]: Yeah, I'll fix it.
It is missing some commas and conjunctions.

File: Code/Framework/AzCore/AzCore/Memory/MallocSchema.cpp (hunk 2 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Memory/MallocSchema.cpp:18]: image

**Comment by lumberyard-employee-dm** [comment]
I actually I have a better idea for these classes.
Instead of trying to use a `#define DEBUG_ALLOCATOR` macro to switch between the debug allocator functionality, we can instead just add a `class HphaSchemaBase<bool>` and `class MallocSchemaBase<bool>` base class that `HphaSchema` and `MallocSchema` inherits from.

The HphaSchemaErrorDetection.cpp test would just instantiate those base class templates using `true` to opt in to the debug functionality.

The HphaSchema would just inherit `HphaSchemaBase<false>` and MallocSchema would just inherit `MallocSchemaBase<false>`

**Comment by lumberyard-employee-dm** [comment]
@burelc-amzn An even simpler solution.
Since the HphaSchemaErrorDetection.cpp file has been deleted with the Allocator Availability work, we are just removing it from the `azcoretest_files.cmake` which prevents it from compiling.

**Comment by lumberyard-employee-dm** [comment]
Hello, anyone?

**Review by byrcolin — APPROVED** [review]

**Comment by lumberyard-employee-dm** [comment]
@burelc-amzn I went back to the solution of using a Base Template, since the HphaSchemaErrorDetectionTests are sticking around as part of the Allocator availability work.

File: Code/Framework/AzCore/AzCore/Memory/HphaSchema.cpp [hunk 3]
```diff
@@ -2599,4 +2697,19 @@ namespace AZ
...
+        return HphaSchemaBase<DebugAllocator>::HpAllocator::MEMORY_GUARD_SIZE;
+    }
+
+    template<bool DebugAllocator>
+    size_t HphaSchemaBase<DebugAllocator>::GetFreeLinkSize()
+    {
+        return sizeof(typename HphaSchemaBase<DebugAllocator>::HpAllocator::free_link);
+    }
+
+    // explicitly instantiate both MallocSchemaBase classes
```
Comment (burelc-amzn) [inline Code/Framework/AzCore/AzCore/Memory/HphaSchema.cpp:2712]:
```suggestion
    // explicitly instantiate both HphaSchemaBase classes
```

File: Code/Framework/AzCore/AzCore/Memory/HphaSchema.cpp [hunk 4]
```diff
@@ -2599,4 +2697,19 @@ namespace AZ
...
+    template class HphaSchemaBase<false>;
+    template class HphaSchemaBase<true>;
```
Comment (burelc-amzn) [inline Code/Framework/AzCore/AzCore/Memory/HphaSchema.cpp:2714]: The only thing that uses `HphachemaBase<true>` is the unit tests. With this change, we're adding 2 copies of the HPHA allocator code to the `AzCore` static library. What impact does that have on our build size? Does it get stripped at link time due to being unused?

File: Code/Framework/AzCore/AzCore/Memory/HphaSchema.cpp [hunk 5]
```diff
@@ -455,7 +450,6 @@ namespace AZ
         };
         static constexpr size_t DEBUG_UNKNOWN_SIZE = (size_t)-1;
 
-#ifdef DEBUG_ALLOCATOR
```
Comment (burelc-amzn) [inline Code/Framework/AzCore/AzCore/Memory/HphaSchema.cpp:458]: This adds 128 bytes to the size of `HpAllocator` when `DebugAllocatorEnable == false`, due to the new `debug_record_map` (line 562), `AZStd::mutex` (line 564), and `size_t[2]` (line 567) members, which are not used when `DebugAllocatorEnable` is `false`. I'd rather see that debug functionality moved out into its own class, so that we're not impacting the `HpAllocator` when `DebugAllocatorEnable` is `false`.

**Review by burelc-amzn — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/Memory/HphaSchema.cpp (hunk 5 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Memory/HphaSchema.cpp:458]: Well there is only 1 HpAllocator in the entire application.
So there 128 bytes total of extra memory used which I think is an acceptable tradeoff.

File: Code/Framework/AzCore/AzCore/Memory/HphaSchema.cpp (hunk 4 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Memory/HphaSchema.cpp]: The size that it adds to the AzCore static library is is about 1 MiB in debug and 600 KiB in profile.

Here is the size of AzCore without the HphaSchema class template split
## Debug
* AzCore.lib - 705,068 KB
* AzCore.pdb - 44,484 KB

## Profile
* AzCore.lib - 267,993 KB
* AzCore.pdb - 46,180 KB

Here is the size of AzCore with the template split
## Debug
* AzCore.lib - 706,012 KB
* AzCore.pdb - 44,484 KB

## Profile
* AzCore.lib - 268,596 KB
* AzCore.pdb - 46,180 KB

At link time the symbol is stripped from all shared libraries and executables except for libAzCore.Tests.so
```bash
[user@arch o3de]$ nm -C build/linux_ninja/bin/profile/libAzCoreTestDLL.so | grep HphaSchemaBase 
0000000000023560 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::DeAllocate(void*, unsigned long, unsigned long)
0000000000023600 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::ReAllocate(void*, unsigned long, unsigned long)
0000000000023650 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::AllocationSize(void*)
00000000000236b0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GarbageCollect()
00000000000237d0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetDebugConfig()
00000000000235b0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Resize(void*, unsigned long)
0000000000023800 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Destroy()
00000000000234f0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Allocate(unsigned long, unsigned long, int, char const*, char const*, int, unsigned int)
0000000000023480 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::~SimpleSchemaAllocator()
0000000000023420 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::~SimpleSchemaAllocator()
00000000000236e0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::NumAllocatedBytes() const
0000000000023740 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetMaxAllocationSize() const
00000000000237a0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetUnAllocatedMemory(bool) const
0000000000023770 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetMaxContiguousAllocationSize() const
0000000000023710 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Capacity() const
00000000000d7698 d typeinfo for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
00000000000a910f r typeinfo name for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
00000000000d75b8 d vtable for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
[mcgarrah@arch o3de]$ nm -C build/linux_ninja/bin/profile/libAzCoreTestDLL.so | grep HphaSchema
0000000000023560 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::DeAllocate(void*, unsigned long, unsigned long)
0000000000023600 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::ReAllocate(void*, unsigned long, unsigned long)
0000000000023650 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::AllocationSize(void*)
00000000000236b0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GarbageCollect()
00000000000237d0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetDebugConfig()
00000000000235b0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Resize(void*, unsigned long)
0000000000023800 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Destroy()
00000000000234f0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Allocate(unsigned long, unsigned long, int, char const*, char const*, int, unsigned int)
0000000000023480 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::~SimpleSchemaAllocator()
0000000000023420 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::~SimpleSchemaAllocator()
00000000000236e0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::NumAllocatedBytes() const
0000000000023740 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetMaxAllocationSize() const
00000000000237a0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetUnAllocatedMemory(bool) const
0000000000023770 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetMaxContiguousAllocationSize() const
0000000000023710 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Capacity() const
00000000000d7698 d typeinfo for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
00000000000a910f r typeinfo name for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
00000000000d75b8 d vtable for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
```

**Review by burelc-amzn — APPROVED** [review]
